parser/src/token.rs

```diff
@@ -516,6 +548,47 @@ impl<'input> Iterator for Tokenizer<'input> {
     }
 }
 
+/// Converts partial hex literal (i.e. part after `0x` or `-0x`) to 64 bit signed integer.
+///
+/// This is basically a copy and adaptation of `std::num::from_str_radix`.
+fn i64_from_hex(hex: &str, is_positive: bool) -> Result<i64, Error> {
+    let radix = 16;
+    let digits = hex.as_bytes();
+    let mut result = 0i64;
+    if is_positive {
+        for &c in digits {
+            let x = match (c as char).to_digit(radix) {
+                Some(x) => x,
+                None => return Err(HexLiteralInvalid),
+            };
+            result = match result.checked_mul(radix as i64) {
+                Some(result) => result,
+                None => return Err(HexLiteralOverflow),
+            };
+            result = match result.checked_add(x as i64) {
+                Some(result) => result,
+                None => return Err(HexLiteralOverflow),
+            };
+        }
+    } else {
+        for &c in digits {
+            let x = match (c as char).to_digit(radix) {
+                Some(x) => x,
+                None => return Err(HexLiteralInvalid),
+            };
+            result = match result.checked_mul(radix as i64) {
```

Comment:
nit: Could write the option matches as `result.checked_mul(radix as i64).ok_or(HexLiteralInvalid)?`